### PR TITLE
Custom import reordering

### LINKED ISSUE #4693: Support custom "project local" crate group in `group_imports`
:clipboard: **Description**

For the sake of example, lets say a project is composed of the following crates: `coolapp-utils`, `coolapp-backend`, and `coolapp-frontend`

It would be great if `group_imports` supported a custom grouping between "external" crates (from crates.io) and "project local" crates:

```rust
// somewhere in `coolapp_frontend`

use alloc::alloc::Layout;
use core::f32;
use std::sync::Arc;

use chrono::Utc;
use uuid::Uuid;
use log::Level;

// At the moment, `group_imports = StdExternalCrate` fuses this group with the group above.
use coolapp_utils::frobnicate;
use coolapp_backend::api::ApiVersion;

use super::schema::{Context, Payload};
use super::update::convert_publish_payload;
use crate::models::Event;
```

Note that different projects have different notions of what constitutes a "project local" crate. Some examples I've come across in the past have included:

- workspace local crates
- crates that share a prefix, but are pulled from crates.io
- crates imported via a local relative path
- etc...

Since there wouldn't be any way to automatically determine which crates fall into which category, one reasonable implementation might be to modify the current "External" crate grouping to support being arbitrarily divided into two (or more?) groups.

While this wouldn't be a \*perfect\* solution (after all, users could still inadvertently import "project local" crates above "external" crates), it would be a step in the right direction, and offer a much needed bit of flexibility to `group_imports`.

### PR DESCRIPTION
**Pull request by l4l**
TLDR, for a given config:

```toml
group_imports = [
    ["$std::*", "proc_macro::*"],
    ["my_crate::*", "crate::*::xyz"],
    ["$crate::*"],
]
```

The following order would be set:

```rust
use proc_macro::Span;
use std::rc::Rc;

use crate::abc::xyz;
use my_crate::a::B;
use my_crate::A;

use self::X;
use super::Y;
use crate::Z;
```

This PR also contains three auxiliary patches for proc-macro `#[config_type]`:

- 0a462a0 allows one to use `#[config_type(skip_derive(Debug, ...))]`, also `impl Copy` removed from most of types where unneeded;
- 1d3e43e is for deriving `serde::{Serialize, Deserialize}`;
- b195129 `impl Display` for singled-fielded enums.

These are not really mandatory but simplify the final commit itself. Probably the proc macro could be refactored further but I think it shall be enough for now.

Implementation notes:

- toml does not support fielded enums at all, that's why manual `impl Deserialize for GroupImportsTactic` is required (https://github.com/toml-rs/toml-rs/issues/302 https://github.com/toml-rs/toml/issues/364);
- don't really want to introduce an additional crate for wildcards so regex is basically used with a hack: `escape(user_pattern).replace("\\*", ".*")`, performance should not be concern here anyway;
- `group_imports` function (i.e for `StdExternalCrate`) could be re-implemented with `group_imports_custom` but I would rather prefer leaving it separate;
- `FromStr` is not implemented for `GroupImportsTactic::Wildcards`;
- `$crate` alias doesn't mean usage of macro-specific `$crate` import.

Resolves #4693 #4788 #5550

### CODE REVIEW

File: Configurations.md [hunk 1]
```diff
@@ -2294,6 +2294,39 @@ use std::sync::Arc;
...
+group_imports = [
+    ["$std::*", "proc_macro::*"],
+    ["my_crate::*", "crate::*::xyz"],
+    ["$crate::*"],
+]
+```
+
+The following order would be set:
+
+```rust
```
Comment (l4l) [inline Configurations.md:2317]: Not really sure how to deal with this one. I could just remove rust syntax so the config test passes, or perhaps there's a better option?

**Comment by l4l** [comment]
Hint: one may build a group for all crates within the workspace with cargo-metadata:

```bash
WORKSPACE_INCLUDES=$(cargo metadata --offline --format-version 1 | jq -r '.workspace_members[]' | awk '{ print $1 }' | sed 's|-|_|' | sed 's|^\(.*\)$|"\1::\*",|' | tr '\n' ' ' | sed 's|, $||')
echo "[${WORKSPACE_INCLUDES}]"
```

which for example for [Rocket](https://github.com/SergioBenitez/Rocket) produces this:

```toml
["rocket::*", "rocket_codegen::*", "rocket_http::*", "rocket_db_pools_codegen::*", "rocket_db_pools::*", "rocket_sync_db_pools_codegen::*", "rocket_sync_db_pools::*", "rocket_dyn_templates::*", "rocket_guide_tests::*"]
```

File: Configurations.md [hunk 2]
```diff
@@ -2294,6 +2294,39 @@ use std::sync::Arc;
...
+Discard existing import groups, and create groups as specified by the wildcarded list.
+Handy aliases are supported:
+
+- `$std` prefix is an alias for standard library (i.e `std`, `core`, `alloc`);
+- `$crate` prefix is an alias for crate-local modules (i.e `self`, `crate`, `super`).
+
+For a given config:
+
+```toml
+group_imports = [
```
Comment (PatchMixolydic) [inline Configurations.md:2308]: It might be handy to document how one could create a group for any crate not already in a group, like in `StdExternalCrate`. Based on the tests for this, I think this might suffice:
```toml
group_imports = [
    ["$std::*", "proc_macro::*"],
    # external crates
    ["*"],
    ["my_crate::*", "crate::*::xyz"],
    ["$crate::*"],
]
```

Comment (l4l) [reply Configurations.md:2308]: Current implementation won't work for your example as you might expect. The groups are checked for a match consequently thus all that doesn't match the first group goes to the second with the last two being always unused.

I don't really think changing implementation for supporting this single case for external crates is a good idea, nor handling myriad of edge-cases for a general solution. But perhaps a note about order would be handy.

**Comment by RicoGit** [comment]
Wow, this is exactly what we need! Is there any updates about this?)

File: Configurations.md (hunk 2 above)
Comment (utkarshgupta137) [reply Configurations.md:2308]: I've raised a PR in your repo for adding custom fallback groups: https://github.com/l4l/rustfmt/pull/1. I think the approach is simple enough that it can be included with this feature. Without this, we're limited to StdWorkspaceCrateExternal (we can add regexs for all external crates, but that's not practical). I don't see myself using this feature (group_imports) if I can't do StdExternalWorkspaceCrate.

Comment (aldanor) [reply Configurations.md:2308]: Agreed with @PatchMixolydic and @utkarshgupta137, being able to specify where "everything else" goes in a simple way without having to jump through hoops seems incredibly useful.

File: Configurations.md (hunk 1 above)
Comment (ytmimi) [reply Configurations.md:2317]: I think you'll need to implement `FromStr` for `GroupImportsTactic`. You can take a look at the config documentation added by  https://github.com/rust-lang/rustfmt/pull/5347, as an example.

Comment (utkarshgupta137) [reply Configurations.md:2317]: @l4l Will you be able to make these changes or should I create a new PR?

Comment (l4l) [reply Configurations.md:2317]: Was having a vacation with no internet. Hopefully will update the PR within this weekend

File: Configurations.md (hunk 2 above)
Comment (l4l) [reply Configurations.md]: Added @utkarshgupta137 patch with an addition of that group being restricted to a single occurrence. Thanks for the opinions and the PR

**Comment by utkarshgupta137** [comment]
@l4l I think you need to rebase your PR to get the tests working.

**Comment by utkarshgupta137** [comment]
Found a failing test case:

conf:
```
group_imports = [
    ["$std::*", "proc_macro::*"],
    ["*"],
    ["workspace*::*"],
    ["$crate::*"],
]
```

Input:
```rust
use std::fmt;

use serde::de;

use workspace_p1::m1;
use workspace_p2::m2;
use workspace_p3 as p3;
use workspace_p4 as p4;
```

Expected:
```rust
use std::fmt;

use serde::de;

use workspace_p1::m1;
use workspace_p2::m2;
use {workspace_p3 as p3, workspace_p4 as p4};
```

Actual:
```rust
use std::fmt;

use serde::de;
use {workspace_p3 as p3, workspace_p4 as p4};

use workspace_p1::m1;
use workspace_p2::m2;
```

**Comment by utkarshgupta137** [comment]
Will probably be useful to add a $pub which covers `pub(crate)` & `pub` uses.

**Comment by l4l** [comment]
For me it works as expected, it's not expected to match `xyz` with `xyz::*` (only with `xyz` or `xyz*`). However the case for the following conf:

```toml
group_imports = [
    ["$std::*", "proc_macro::*"],
    ["*"],
    ["workspace*"],
]
```

works indeed in a quite unexpected way (there's the same result). There's no simple way to split includes, reorder and then somehow "preserve" original granularity. I don't think this behavior need to be changed. The actual problem here is missing `import_granularity` (e.g `Crate`) option which places all of the `workspace_p*` modules in the third block.

**Comment by tinct-martini** [comment]
Is there any chance to get this reviewed and merged?
Seems like this has been stuck in limbo for no good reason when it's an excellent feature a lot of us are waiting for 😃

**Comment by PatchMixolydic** [comment]
Experience report: I tried this option in a fork of `rustfmt` in a decently-sized personal project (~18 kLoC) with the following config:
```toml
# ...
group_imports = [["*"], ["$crate::*"]]
# ...
```

I only ran into one surprising "issue": a local module import was treated as an external crate import:
```diff
  mod register;
  mod view;

  use glam::{const_vec3, Vec2, Vec3};
  use mountain_river::high_scores::{HighScore, HighScores, HIGH_SCORES_PER_GROUP};
+ pub use register::RegisterHighScore;
  use std::{array, borrow::Cow};
  use triplicata::{
      // ...
  };
-
- pub use register::RegisterHighScore;
  pub use view::ViewHighScores;
```

Since this feature works by matching on import paths, this doesn't really seem like a bug. This was easily fixed by prepending `self::` to both local imports.

Also, note that it moved a reexport into a group containing normal imports. This is slightly unfortunate, but I believe that this is an acceptable price to pay for consistency.

Otherwise, everything seems to have gone quite smoothly!

**Comment by l4l** [comment]
It seems this PR got forgotten, so kind reminder @ytmimi @calebcartwright 

If anything else you needed from me or have any concerns feel free to ping. If it's better to split-out the proc-macro part that's also not a problem.

**Comment by ytmimi** [comment]
@l4l thanks for reaching out! I wouldn't say this has been forgotten, it's just been put on the back burner while we tend to some higher priority items. We spent a lot of time getting ready for `if-let` support, and we just recently released some experimental support for `let-chains`. The next big ticket item for rustfmt is implementing [style_edition](https://rust-lang.github.io/rfcs/3338-style-evolution.html) to give the project more flexibility moving forward. I hope you understand and I appreciate your patience on this.

In the interim, maybe you can add additional tests to see how this new option interacts with other import options like `imports_granularity`. If there are any edge cases you can think of having good tests for those would also be helpful. While Caleb and I can't spend time reviewing this right now it doesn't mean that this PR has to go unreviewed. Thank you to everyone who's already chimed in and shared their thoughts. I'd sure welcome reviews from anyone who's interested in seeing this land continue to provide feedback on how this PR could be improved or how the implementation could be simplified. It would definitely be a great help for when Caleb and I are able to revisit this.

I'd also like to add that I understand this is important to you and many others, and I personally think it looks like a cool feature!

**Comment by tgross35** [comment]
Some random drive-by notes:

1. It would be a nice convenience if a string is interchangeable with a single-item list in the config. That is, `group_imports = ["$std", "*"]` would act the same as `group_imports = [["$std"], ["*"]]`
2. I think regex may be better to use here rather than wildcards because it allows e.g.  `myproject_(macros|cli)::.*` or `rustc_.*`. In which case, I think it is better to suggest using `.*` directly rather than doing `replace("*", ".*")`, which would turn `.*` in to `..*`
3. How is the fallback supposed to work? In the opening example:

```toml
group_imports = [
    ["$std::*", "proc_macro::*"],
    ["*"],
    ["my_crate::*", "crate::*::xyz"],
    ["$crate::*"],
]
```

I haven't looked at the algorithm but I assume it's trying to match from most specific to most general somehow. It would be more straightforward IMO to always match first to last, and then add a meta `$other` or `$fallback` matcher where everything goes if it isn't picked up by another group. E.g. the above could be written as the following (including my other suggestions):

```toml
group_imports = [
    ["$std::.*", "proc_macro::.*"],
    "$other",
    ["my_crate::.*", "crate::.*::xyz"],
    ["$crate::.*"],
]
```

**Comment by DXist** [comment]
Very useful PR!

It's possible to use [globset][(https://github.com/BurntSushi/ripgrep/tree/master/crates/globset](https://docs.rs/globset/latest/globset/) as a frontend to regex engine.

It supports "or" globs (`myproject_{macros,cli}::*`), cc @tgross35.
